src/Disks/ObjectStorages/MetadataStorageFromPlainObjectStorageOperations.cpp

```diff
@@ -20,38 +21,57 @@ namespace
 
 constexpr auto PREFIX_PATH_FILE_NAME = "prefix.path";
 
+ObjectStorageKey createMetadataObjectKey(const std::string & key_prefix, const std::string & metadata_key_prefix)
+{
+    auto prefix = std::filesystem::path(metadata_key_prefix) / key_prefix;
+    return ObjectStorageKey::createAsRelative(prefix.string(), PREFIX_PATH_FILE_NAME);
+}
 }
 
 MetadataStorageFromPlainObjectStorageCreateDirectoryOperation::MetadataStorageFromPlainObjectStorageCreateDirectoryOperation(
-    std::filesystem::path && path_,
-    std::string && key_prefix_,
-    MetadataStorageFromPlainObjectStorage::PathMap & path_map_,
-    ObjectStoragePtr object_storage_)
-    : path(std::move(path_)), key_prefix(key_prefix_), path_map(path_map_), object_storage(object_storage_)
+    std::filesystem::path && path_, InMemoryPathMap & path_map_, ObjectStoragePtr object_storage_, const std::string & metadata_key_prefix_)
+    : path(std::move(path_))
+    , path_map(path_map_)
+    , object_storage(object_storage_)
+    , metadata_key_prefix(metadata_key_prefix_)
+    , key_prefix(object_storage->generateObjectKeyPrefixForDirectoryPath(path, "" /* key_prefix */).serialize())
 {
 }
 
 void MetadataStorageFromPlainObjectStorageCreateDirectoryOperation::execute(std::unique_lock<SharedMutex> &)
 {
-    if (path_map.contains(path))
-        return;
+    auto & mutex = path_map.mutex;
+    {
+        std::shared_lock lock(mutex);
+        auto & map = path_map.map;
+        if (map.contains(path.parent_path()))
+            return;
+    }
 
-    LOG_TRACE(getLogger("MetadataStorageFromPlainObjectStorageCreateDirectoryOperation"), "Creating metadata for directory '{}'", path);
+    auto metadata_object_key = createMetadataObjectKey(key_prefix, metadata_key_prefix);
 
-    auto object_key = ObjectStorageKey::createAsRelative(key_prefix, PREFIX_PATH_FILE_NAME);
+    LOG_TRACE(
+        getLogger("MetadataStorageFromPlainObjectStorageCreateDirectoryOperation"),
+        "Creating metadata for directory '{}' with remote path='{}'",
+        path,
+        metadata_object_key.serialize());
 
-    auto object = StoredObject(object_key.serialize(), path / PREFIX_PATH_FILE_NAME);
+    auto metadata_object = StoredObject(metadata_object_key.serialize(), path / PREFIX_PATH_FILE_NAME);
     auto buf = object_storage->writeObject(
-        object,
+        metadata_object,
         WriteMode::Rewrite,
         /* object_attributes */ std::nullopt,
         /* buf_size */ DBMS_DEFAULT_BUFFER_SIZE,
         /* settings */ {});
 
     write_created = true;
 
-    [[maybe_unused]] auto result = path_map.emplace(path, std::move(key_prefix));
-    chassert(result.second);
+    {
+        std::unique_lock lock(mutex);
+        auto & map = path_map.map;
+        [[maybe_unused]] auto result = map.emplace(path.parent_path(), std::move(key_prefix));
```

Comment:
why take `path.parent_path()` here and not just `path`? `path` should already be a directory we want (as this is CreateDirectoryOperation) or I am misunderstanding?

Reply:
```
std::cout << std::filesystem::path("/path/to/file/").parent_path();
"/path/to/file"
```
It's just a weird way to trim `'/'`. I'll make it more explicit, thanks!